### PR TITLE
[model] fix: qwen3-moe-a3b load

### PR DESCRIPTION
**Pull request by ziqi-wlb**
### fix qwen3-moe-a3b load

When running the qwen3-moe-a3b model, I found that the initial loss(17+) was too high. The final location was a bug in the model loading, then the experts were randomly initialized.

- before fix, first-step loss=17
image

- solution:
1. The loading problem is mainly due to the mismatch between the model definition and the checkpoint format: VeOmni's Qwen3MoeExperts uses a single stacked parameter `torch.nn.Parameter(self.num_experts, ...)`, while the hf-checkpoint uses independent expert parameters `experts.0.gate_proj.weight`, `experts.1.gate_proj.weight`, etc.
2. Loading speed optimization: There are multiple experts, for example, qwen3-moe-a3b has 128 experts. Searching and loading them individually each time will be extremely slow. The cache mechanism is used to avoid repeated module searches, and batch loading is used to merge multiple experts into a single loading operation.

- after fix, first-step loss=2.8
image

- Reproduction script:
```sh

MODEL_PATH=/cpfs/user/weishi/models/Qwen3-30B-A3B
SAVE_PATH=Qwen3-30B-A3B
CONFIG=configs/pretrain/qwen3-moe.yaml

DATA_PATH=/prodcpfs/user/weishi/data/fineweb/sample/10BT/
DATA_SET_TYPE=mapping

bash train.sh tasks/train_torch.py $CONFIG \
    --model.model_path ${MODEL_PATH} \
    --data.train_path $DATA_PATH \
    --data.num_workers 32 \
    --data.datasets_type $DATA_SET_TYPE \
    --train.global_batch_size 128 \
    --train.lr 5e-7 \
    --train.ulysses_parallel_size 1 \
    --train.save_steps 10 \
    --train.max_steps 20 \
    --train.ckpt_manager dcp \
    --train.output_dir ${SAVE_PATH}
```

### CODE REVIEW

**Comment by TueVNguyen** [comment]
Did you run scripts/moe_merge.py first?

**Comment by ziqi-wlb** [comment]
> Did you run scripts/moe_merge.py first?

OK, `scripts/moe_merge.py` is not executed (it is not mentioned in the document, I did not see this file~). From the perspective of ease of use, it should be insensitive to users. Reducing manual operations is relatively more elegant?

**Comment by Fazziekey** [comment]
see https://github.com/ByteDance-Seed/VeOmni/blob/main/scripts/moe_ckpt_merge/moe_merge.py

**Comment by ziqi-wlb** [comment]
> see https://github.com/ByteDance-Seed/VeOmni/blob/main/scripts/moe_ckpt_merge/moe_merge.py

We use fsdp to train models because it is simpler to use than megatron-core, so the burden of manual operation should be reduced as much as possible, and there are no errors or loading failures that would cause people to step into the trap~ (For example, the early megatron-core had to rely on different TP/PP ckpt splits before training, which was always complained about, but was later optimized)

**Comment by Luosuu** [comment]
Hi @ziqi-wlb , thank you very much for your efforts and the PR! This is great work from the user experience.

However, we believe a more reasonable way of storing experts weight should be the "merged" instead of the current style of Qwen3-MoE HuggingFace checkpoints. Personally, I believe this is a design mistake on their side. Instead of trying to catching their mistakes, we prefer to have a dedicated script to "fix" their checkpoints first. 

From another perspective,  we should not assume all other models have the same way of storing expert weights with the current Qwen3-MoE hf checkpoints. We do not write different loaders / dispatchers for different models inside the library itself, which can be pretty messy from the view of maintainability.

By holding a simple assumption of the "merged" expert weight form, we can 
1. keep the core codebase easy to read
2. make model loading part consistent to different MoE models
3. more maintainable by providing dedicated scripts to fix model weights if they are not in the way we expected

Hopefully this make sense to you

**Comment by ziqi-wlb** [comment]
Ok, I'll close this PR.
